migrations/src/main/scala/za/co/absa/enceladus/migrations/framework/migration/CollectionMigration.scala

```diff
@@ -0,0 +1,293 @@
+    *
+    * @param collectionName A collection name to be added
+    */
+  def dropCollection(collectionName: String): Unit = {
+    if (collectionsToDrop.contains(collectionName)) {
+      throw new IllegalArgumentException(s"Collection '$collectionName' is already in the removal list.")
+    }
+    if (collectionsToCreate.contains(collectionName)) {
+      throw new IllegalArgumentException(s"Cannot both add and remove '$collectionName' as a migration step.")
+    }
+    collectionsToDrop += collectionName
+  }
+
+  /**
+    * This method is used by derived classes to rename a collection as a step of migration process.
+    *
+    * @param oldName A collection to be renamed
+    * @param newName A new name for the collection
+    */
+  def renameCollection(oldName: String, newName: String): Unit = {
+    if (collectionsToDrop.contains(oldName)) {
+      throw new IllegalArgumentException(s"Collection '$oldName' is in the removal list.")
+    }
+    if (collectionsToCreate.contains(oldName)) {
+      throw new IllegalArgumentException(s"Collection '$oldName' is in the list of new collections. Cannot rename it.")
+    }
+    collectionsToRename += oldName -> newName
+  }
+
+  /**
+    * This method is used by derived classes to create an index on a collection.
+    *
+    * @param collectionName A collection for setting up an index
+    * @param fields         A list of fields that the index should contain
+    */
+  def createIndex(collectionName: String, fields: Seq[String]): Unit = {
+    indexesToCreate.append((collectionName, fields))
+  }
+
+  /**
+    * This method is used by derived classes to drop an index on a collection.
+    *
+    * @param collectionName A collection for dropping up an index
+    * @param fields         A list of fields that the index should contain
+    */
+  def dropIndex(collectionName: String, fields: Seq[String]): Unit = {
+    indexesToDrop.append((collectionName, fields))
+  }
+
+  /** Returns a list of collections to be added during the migration */
+  def getCollectionsToAdd: List[String] = collectionsToCreate.toList
+
+  /** Returns a list of collections to be removed during the migration */
+  def getCollectionsToRemove: List[String] = collectionsToDrop.toList
+
+  /** Returns a list of collections to be removed during the migration */
+  def getCollectionsToRename: List[(String, String)] = collectionsToRename.toList
+
+  /** Returns a list of indexes to be added during the migration */
+  def getIndexesToAdd: List[Index] = indexesToCreate.toList
+
+  /** Returns a list of indexes to be removed during the migration */
+  def getIndexesToRemove: List[Index] = indexesToDrop.toList
+
+  /**
+    * If a migration adds or removes collections it should provide a new list of collections based
+    * on the list of collections available for he previous version of a database.
+    */
+  override def applyCollectionChanges(collections: List[String]): List[String] = {
+    var newCollections = collections
```

Comment:
If we used `Set` with `++` and `--` we could omit the var. But that would not keep the order.

Reply:
It is better to keep order. Will switch to ArrayBuffer to remove the var since we remove items from the middle and insert at the end.